### PR TITLE
[BugFix] concurrent-safe issue of cached cpu cores

### PR DESCRIPTION
**Pull request by murphyatwork**
Fixes #issue

Changes:
1. Print the stats when updating
2. Use a pessimistic lock when updating cached values

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Review by satanson — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 13 / 13 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/system/BackendCoreStat.java|13|13|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/system/BackendCoreStat.java [hunk 1]
```diff
@@ -16,18 +16,27 @@
...
 
-    private static ConcurrentHashMap<Long, Integer> numOfHardwareCoresPerBe = new ConcurrentHashMap<>();
+    private static final ConcurrentHashMap<Long, Integer> numOfHardwareCoresPerBe = new ConcurrentHashMap<>();
     private static AtomicInteger cachedAvgNumOfHardwareCores = new AtomicInteger(-1);
+    private static final ReentrantLock lock = new ReentrantLock();
 
     public static void setNumOfHardwareCoresOfBe(long be, int numOfCores) {
         Integer previous = numOfHardwareCoresPerBe.put(be, numOfCores);
+        LOG.info("set numOfHardwareCoresOfBe of be({}) to {}", be, numOfCores);
+        LOG.info("current cpuCores stats: {}", numOfHardwareCoresPerBe);
```
Comment (silverbullet233) [inline fe/fe-core/src/main/java/com/starrocks/system/BackendCoreStat.java:39]: better to merge these two logs into one?

File: fe/fe-core/src/main/java/com/starrocks/system/BackendCoreStat.java [hunk 2]
```diff
@@ -60,26 +69,28 @@ public static int getCoresOfBe(long beId) {
...
+            lock.lock();
+            if (numOfHardwareCoresPerBe.isEmpty()) {
+                return DEFAULT_CORES_OF_BE;
+            }
+            int avg = numOfHardwareCoresPerBe.values().stream().reduce(Integer::sum).orElse(0)
+                    / numOfHardwareCoresPerBe.size();
+            avg = Math.max(DEFAULT_CORES_OF_BE, avg);
+            cachedAvgNumOfHardwareCores.set(avg);
+            LOG.info("update avgNumOfHardwareCoresOfBe to {}", avg);
+            LOG.info("current cpuCores stats: {}", numOfHardwareCoresPerBe);
```
Comment (silverbullet233) [inline fe/fe-core/src/main/java/com/starrocks/system/BackendCoreStat.java:89]: same as above

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 12 / 12 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/system/BackendCoreStat.java|12|12|100.00%|[]|

**Review by ZiheLiu — DISMISSED** [review]

**Review by silverbullet233 — DISMISSED** [review]

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 12 / 12 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/system/BackendCoreStat.java|12|12|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 12 / 12 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/system/BackendCoreStat.java|12|12|100.00%|[]|

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 12 / 12 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/system/BackendCoreStat.java|12|12|100.00%|[]|

**Review by satanson — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**